File: packages/dsa-py/src/structures/stack.py
"""
Stack — LIFO (Last In, First Out)

Complexity:
  push   O(1)
  pop    O(1)
  peek   O(1)
"""
from typing import Generic, TypeVar

T = TypeVar("T")


class Stack(Generic[T]):
    def __init__(self) -> None:
        self._items: list[T] = []

    def push(self, item: T) -> None:
        """Agrega un elemento al tope. O(1)"""
        self._items.append(item)

    def pop(self) -> T | None:
        """Retira y devuelve el elemento del tope. O(1)"""
        return self._items.pop() if self._items else None

    def peek(self) -> T | None:
        """Devuelve el tope sin retirarlo. O(1)"""
        return self._items[-1] if self._items else None

    @property
    def size(self) -> int:
        return len(self._items)

    @property
    def is_empty(self) -> bool:
        return len(self._items) == 0

    def clear(self) -> None:
        self._items.clear()

File: packages/dsa-ts/src/structures/Stack.ts
/**
 * Stack — LIFO (Last In, First Out)
 *
 * Complexity:
 *   push   O(1)
 *   pop    O(1)
 *   peek   O(1)
 *   size   O(1)
 */
export class Stack<T> {
  private readonly items: T[] = [];

  /** Agrega un elemento al tope. O(1) */
  push(item: T): void {
    this.items.push(item);
  }

  /** Retira y devuelve el elemento del tope. O(1) */
  pop(): T | undefined {
    return this.items.pop();
  }

  /** Devuelve el tope sin retirarlo. O(1) */
  peek(): T | undefined {
    return this.items[this.items.length - 1];
  }

  /** Cantidad de elementos. O(1) */
  get size(): number {
    return this.items.length;
  }

  /** True si no hay elementos. O(1) */
  get isEmpty(): boolean {
    return this.items.length === 0;
  }

  /** Elimina todos los elementos. O(1) */
  clear(): void {
    this.items.length = 0;
  }

  /** Itera de tope a fondo. */
  [Symbol.iterator](): Iterator<T> {
    let index = this.items.length - 1;
    const items = this.items;
    return {
      next(): IteratorResult<T> {
        if (index >= 0) {
          return { value: items[index--] as T, done: false };
        }
        return { value: undefined as unknown as T, done: true };
      },
    };
  }
}
